src/index/index.cc

```diff
@@ -34,6 +36,40 @@ LoadConfig(BaseConfig* cfg, const Json& json, knowhere::PARAM_TYPE param_type, c
     return Config::Load(*cfg, json_, param_type, msg);
 }
 
+#ifdef KNOWHERE_WITH_CARDINAL
+template <typename T>
+inline const std::unique_ptr<Handle>
+Index<T>::BuildAsync(const DataSetPtr dataset, const Json& json, const std::chrono::seconds timeout) {
+    auto pool = ThreadPool::GetGlobalBuildThreadPool();
+    auto handle = std::make_unique<Handle>(timeout);
+    handle->Set(pool->push([this, dataset, &json, &handle]() {
+        auto cfg = this->node->CreateConfig();
+        RETURN_IF_ERROR(LoadConfig(cfg.get(), json, knowhere::TRAIN, "Build"));
+
+#if defined(NOT_COMPILE_FOR_SWIG) && !defined(KNOWHERE_WITH_LIGHT)
+        TimeRecorder rc("BuildAsync index ", 2);
+        auto res = this->node->BuildAsync(dataset, std::move(cfg), handle.get());
+        auto time = rc.ElapseFromBegin("done");
+        time *= 0.000001;  // convert to s
+        knowhere_build_latency.Observe(time);
+#else
+        auto res = this->node->BuildAsync(dataset, std::move(cfg), handle.get());
+#endif
+        return res;
+    }));
+    return handle;
+}
+#else
+template <typename T>
+inline const std::unique_ptr<Handle>
+Index<T>::BuildAsync(const DataSetPtr dataset, const Json& json) {
```

Comment:
same. Please switch to `std::shared_ptr<Handle>`

Reply:
Each `async build` call generates only one handle object, and all references are just pointers. Moreover, the called functions are all `const` functions.

Reply:
the reason why is that 
```
    interrupt->Set(pool->push([this, dataset, &json, &interrupt]() {
```

Basically, an object captures a reference to smart pointer. This needs to be implemented as capturing the smart pointer itself, not a reference to it. Thus, `shared_ptr` is needed.

I understand the suggestion, but the reality is that C++ async brings to the territory of extremely nasty bugs. Thus, we need to make very pessimistic assumptions on code usage.